src/types/complex.rs

```diff
@@ -19,118 +19,224 @@ pyobject_native_type!(
+        )
+    )]
     pub fn from_doubles(py: Python<'_>, real: c_double, imag: c_double) -> &PyComplex {
+        Self::from_doubles_bound(py, real, imag).into_gil_ref()
+    }
+    /// Creates a new `PyComplex` from the given real and imaginary values.
+    pub fn from_doubles_bound(
+        py: Python<'_>,
+        real: c_double,
+        imag: c_double,
+    ) -> Bound<'_, PyComplex> {
+        use crate::ffi_ptr_ext::FfiPtrExt;
         unsafe {
-            let ptr = ffi::PyComplex_FromDoubles(real, imag);
-            py.from_owned_ptr(ptr)
+            ffi::PyComplex_FromDoubles(real, imag)
+                .assume_owned(py)
+                .downcast_into_unchecked()
         }
     }
     /// Returns the real part of the complex number.
     pub fn real(&self) -> c_double {
-        unsafe { ffi::PyComplex_RealAsDouble(self.as_ptr()) }
+        self.as_borrowed().real()
     }
     /// Returns the imaginary part of the complex number.
     pub fn imag(&self) -> c_double {
+        self.as_borrowed().imag()
+    }
+}
+
+/// Implementation of functionality for [`PyComplex`].
+///
+/// These methods are defined for the `Bound<'py, PyComplex>` smart pointer, so to use method call
+/// syntax these methods are separated into a trait, because stable Rust does not yet support
+/// `arbitrary_self_types`.
+#[doc(alias = "PyComplex")]
+pub trait PyComplexMethods<'py> {
+    /// Returns the real part of the complex number.
+    fn real(&self) -> c_double;
+    /// Returns the imaginary part of the complex number.
+    fn imag(&self) -> c_double;
+    /// Returns `|self|`.
+    #[cfg(not(any(Py_LIMITED_API, PyPy)))]
+    fn abs(&self) -> c_double;
+    /// Returns `self` raised to the power of `other`.
+    #[cfg(not(any(Py_LIMITED_API, PyPy)))]
+    fn pow(&self, other: &Bound<'py, PyComplex>) -> Bound<'py, PyComplex>;
+}
+
+impl<'py> PyComplexMethods<'py> for Bound<'py, PyComplex> {
+    fn real(&self) -> c_double {
+        unsafe { ffi::PyComplex_RealAsDouble(self.as_ptr()) }
+    }
+
+    fn imag(&self) -> c_double {
         unsafe { ffi::PyComplex_ImagAsDouble(self.as_ptr()) }
     }
+
+    #[cfg(not(any(Py_LIMITED_API, PyPy)))]
+    fn abs(&self) -> c_double {
+        unsafe {
+            let val = (*(self.as_ptr() as *mut ffi::PyComplexObject)).cval;
+            ffi::_Py_c_abs(val)
+        }
+    }
+
+    #[cfg(not(any(Py_LIMITED_API, PyPy)))]
+    fn pow(&self, other: &Bound<'py, PyComplex>) -> Bound<'py, PyComplex> {
+        use crate::ffi_ptr_ext::FfiPtrExt;
+        unsafe {
+            not_limited_impls::complex_operation(
+                self.as_borrowed(),
+                other.as_borrowed(),
+                ffi::_Py_c_pow,
+            )
+            .assume_owned(self.py())
+            .downcast_into_unchecked()
+        }
+    }
 }
 
 #[cfg(not(any(Py_LIMITED_API, PyPy)))]
 mod not_limited_impls {
```

Comment:
I think I have a slight preference to have this `mod not_limited_impls` above the `PyComplexMethods` trait, but it's not a strong opinion if you had a reason for this ordering.

Reply:
I just put it there because this look roughly like the location it was added in the other modules, but I'm happy to move it down if you prefer that.